src/gfn/envs/discrete_ebm.py

```diff
@@ -140,37 +140,43 @@ def is_exit_actions(self, actions: BatchTensor) -> BatchTensor:
         return actions == self.n_actions - 1
 
     def maskless_step(self, states: States, actions: Actions) -> StatesTensor:
-        # First, we select that actions that replace a -1 with a 0
-        mask_0 = actions.actions_tensor < self.ndim
+        # First, we select that actions that replace a -1 with a 0.
+        # Remove singleton dimension for broadcasting. TODO: is this correct?
+        mask_0 = (actions.actions_tensor < self.ndim).squeeze(-1)
         states.states_tensor[mask_0] = states.states_tensor[mask_0].scatter(
-            -1, actions.actions_tensor[mask_0], 0
-        )
-        # Then, we select that actions that replace a -1 with a 1
-        mask_1 = (actions.actions_tensor >= self.ndim) & (
-            actions.actions_tensor < 2 * self.ndim
+            -1, actions.actions_tensor[mask_0], 0  # Set indices to 0.
         )
+        # Then, we select that actions that replace a -1 with a 1.
+        mask_1 = (
+            (actions.actions_tensor >= self.ndim)
+            & (actions.actions_tensor < 2 * self.ndim)
+        ).squeeze(
+            -1
+        )  # Remove singleton dimension for broadcasting.
         states.states_tensor[mask_1] = states.states_tensor[mask_1].scatter(
-            -1, (actions.actions_tensor[mask_1] - self.ndim), 1
+            -1, (actions.actions_tensor[mask_1] - self.ndim), 1  # Set indices to 1.
         )
         return states.states_tensor
 
     def maskless_backward_step(self, states: States, actions: Actions) -> StatesTensor:
         return states.states_tensor.scatter(
-            -1, actions.unsqueeze(-1).fmod(self.ndim), -1
+            -1,
+            actions.actions_tensor.fmod(self.ndim),
+            -1,  # TODO: Why is fmod required?
```

Comment:
For this environment, states are n-dimensional vectors. s0 is empty (empty corresponds to -1): s0 = [-1, -1, ...., -1], and each action corresponds to replacing a -1 with either 0 or 1.

 Action i in [0, ndim - 1] corresponds to replacing s[i] with 0.
 Action i in [ndim, 2 * ndim - 1] corresponds to replacing s[i - ndim] with 1.

So when taking a backward action, we are essentially asking the question "what index should be set back to -1?"

Thus the fmod